### PR TITLE
Quorum store cache

### PR DESCRIPTION
**Pull request by zjma**
### Test Plan

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3422)

### CODE REVIEW

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs [hunk 1]
```diff
@@ -4,6 +4,11 @@
 mod storage_wrapper;
 mod vm_wrapper;
 
+use aptos_crypto::hash::DefaultHasher;
+use bcs::to_bytes;
+use concurrent_lru::sharded::LruCache;
```
Comment (zjma) [inline aptos-move/aptos-vm/src/parallel_executor/mod.rs:9]: this lib is ok but not optimal. We can take some time to implement our own if needed.
- The number of shard is fixed (16, as of v2.0).
- The lib provides map-like cache with API like `get_or_insert(key, func-to-gen-value-if-key-absent) -> actual-value`, while we need something like "insert and return if it's a cache hit". The best we can do is to insert with a `nonce`, then claim it's a cache miss if `actual-value==nonce`. This makes false cache miss possible.

**Comment by alinush** [comment]
Zhoujun, can you describe what this PR is for in the Description above?

File: aptos-move/parallel-executor/src/executor.rs [hunk 2]
```diff
@@ -100,29 +91,24 @@ impl<'a, K: ModulePath + PartialOrd + Send + Clone + Hash + Eq, V: Send + Sync>
...
     E: ExecutorTask<T = T>,
 {
-    /// The caller needs to ensure that concurrency_level > 1 (0 is illegal and 1 should
-    /// be handled by sequential execution) and that concurrency_level <= num_cpus.
-    pub fn new(concurrency_level: usize) -> Self {
-        assert!(
-            concurrency_level > 1 && concurrency_level <= num_cpus::get(),
-            "Parallel execution concurrency level {} should be between 2 and number of CPUs",
-            concurrency_level
-        );
```
Comment (alinush) [inline aptos-move/parallel-executor/src/executor.rs:122]: Why remove this assert?

File: crates/aptos-crypto/src/hash.rs [hunk 3]
```diff
@@ -144,6 +144,11 @@ impl HashValue {
             .map(Self::new)
     }
 
+    /// Get byte array representation.
+    pub fn get_bytes(&self) -> [u8; HashValue::LENGTH] {
```
Comment (alinush) [inline crates/aptos-crypto/src/hash.rs:148]: If you are cloning it, you might as well just use `HashValue::to_vec()` instead of defining a new method.

Comment (alinush) [reply crates/aptos-crypto/src/hash.rs:148]: See https://github.com/aptos-labs/aptos-core/blob/92f8b659094f2643a33a0d98f74b41dd676144f5/crates/aptos-crypto/src/hash.rs#L148

**Comment by alinush** [comment]
@sasha8 @gelash @bchocho can you add yourselves as code-owners in `aptos-core/CODEOWNERS` in the proper repos, so you get automatically notified about relevant PRs?

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs (hunk 1 above)
Comment (gelash) [reply aptos-move/aptos-vm/src/parallel_executor/mod.rs:9]: I think I'd implement our own, should be fun and relatively small implementation where we can control most things (and testing is also not too complex). The crate is also not super-popular.

File: aptos-move/parallel-executor/src/proptest_types/bencher.rs [hunk 4]
```diff
@@ -103,9 +103,14 @@ where
     }
 
     pub(crate) fn run(self) {
+        let thread_pool = rayon::ThreadPoolBuilder::new()
```
Comment (gelash) [inline aptos-move/parallel-executor/src/proptest_types/bencher.rs:106]: I don't think we ever looked into fake bencher numbers too much, but still we probably shouldn't be creating the threadpool in run here. For other tests it's probably fine.

File: aptos-move/aptos-vm/src/parallel_executor/mod.rs [hunk 5]
```diff
@@ -58,6 +64,47 @@ impl PTransactionOutput for AptosTransactionOutput {
...
+        let mut hasher = DefaultHasher::new(b"CacheTesting");
+        hasher.update(&bytes);
+        hasher.finish().get_bytes()
+    }
+
+    /// return whether the item exists (cache hit).
+    /// NOTE: false cache miss is possible.
+    pub fn insert(&self, item: &Transaction) -> bool {
+        let nonce = thread_rng().gen::<u64>();
+        let entry = self.state.get_or_init(self.get_key(item), 1, |_e| nonce);
```
Comment (gelash) [inline aptos-move/aptos-vm/src/parallel_executor/mod.rs:97]: oh this nonce business :-)
definitely doesn't look like a good library, given the APIs are so limited, shards fixed.

File: aptos-move/parallel-executor/src/executor.rs (hunk 2 above)
Comment (zjma) [reply aptos-move/parallel-executor/src/executor.rs:122]: clearly i was drunk

**Comment by davidiw** [comment]
is this pr still alive? closing as it looks stale.

**Pull request closed without merging**